tests/queries/0_stateless/01576_alias_column_rewrite.sql

```diff
@@ -0,0 +1,89 @@
+
+INSERT INTO table_with_alias_column(timestamp, value) SELECT toDateTime('2020-01-02 12:00:00'), 1 FROM numbers(10);
+
+INSERT INTO table_with_alias_column(timestamp, value) SELECT toDateTime('2020-01-03 12:00:00'), 1 FROM numbers(10);
+
+
+SELECT 'test-partition-prune';
+
+SELECT COUNT() = 10 FROM table_with_alias_column WHERE day = '2020-01-01' SETTINGS max_rows_to_read = 10;
+SELECT t = '2020-01-03' FROM (SELECT day as t FROM table_with_alias_column WHERE t = '2020-01-03' GROUP BY t SETTINGS max_rows_to_read = 10);
+SELECT COUNT() = 10 FROM table_with_alias_column WHERE day = '2020-01-01' UNION ALL select 1 from numbers(1) SETTINGS max_rows_to_read = 11;
+SELECT  COUNT() = 0 FROM (SELECT  toDate('2019-01-01') as  day, day as t   FROM table_with_alias_column PREWHERE t = '2020-01-03'  WHERE t  = '2020-01-03' GROUP BY t );
+
+SELECT 'test-join';
+
+SELECT day = '2020-01-03'
+FROM
+(
+ SELECT toDate('2020-01-03') AS day
+ FROM numbers(1)
+) AS a
+INNER JOIN
+(
+ SELECT day
+ FROM table_with_alias_column
+ WHERE day = '2020-01-03'
+ GROUP BY day SETTINGS max_rows_to_read = 11
+) AS b ON a.day = b.day;
+
+SELECT day = '2020-01-01'
+FROM
+(
+ SELECT day
+ FROM table_with_alias_column
+ WHERE day = '2020-01-01'
+ GROUP BY day SETTINGS max_rows_to_read = 11
+) AS a
+INNER JOIN
+(
+ SELECT toDate('2020-01-01') AS day
+ FROM numbers(1)
+) AS b ON a.day = b.day;
+
+
+SELECT 'alias2alias';
+SELECT COUNT() = 10 FROM table_with_alias_column WHERE day1 = '2020-01-02' SETTINGS max_rows_to_read = 10;
+SELECT t = '2020-01-03' FROM (SELECT day1 as t FROM table_with_alias_column WHERE t = '2020-01-03' GROUP BY t SETTINGS max_rows_to_read = 10);
+SELECT t = '2020-01-03' FROM (SELECT day2 as t FROM table_with_alias_column WHERE t = '2020-01-03' GROUP BY t SETTINGS max_rows_to_read = 10);
+SELECT COUNT() = 10 FROM table_with_alias_column WHERE day1 = '2020-01-03' UNION ALL select 1 from numbers(1) SETTINGS max_rows_to_read = 11;
+SELECT  COUNT() = 0 FROM (SELECT  toDate('2019-01-01') as  day1, day1 as t   FROM table_with_alias_column PREWHERE t = '2020-01-03'  WHERE t  = '2020-01-03' GROUP BY t );
+SELECT day1 = '2020-01-04' FROM table_with_alias_column PREWHERE day1 = '2020-01-04'  WHERE day1 = '2020-01-04' GROUP BY day1 SETTINGS max_rows_to_read = 10;
+
+DROP TABLE table_with_alias_column;
+
+
+SELECT 'second_index';
+
+DROP TABLE IF EXISTS test_index;
+CREATE TABLE test_index
+(
+    `key_string` String,
+    `key_uint32` ALIAS toUInt32(key_string),
+    INDEX idx toUInt32(key_string) TYPE set(0) GRANULARITY 1
+)
+ENGINE = MergeTree
+PARTITION BY tuple()
+PRIMARY KEY tuple()
+ORDER BY key_string SETTINGS index_granularity = 1;
+
+INSERT INTO test_index SELECT * FROM numbers(10);
+SELECT COUNT() == 1 FROM test_index WHERE key_uint32 = 1 SETTINGS max_rows_to_read = 10;
+SELECT COUNT() == 1 FROM test_index WHERE toUInt32(key_string) = 1 SETTINGS max_rows_to_read = 10;
+DROP TABLE IF EXISTS test_index;
```

Comment:
Correct me if I'm wrong but this will not work for:
- optimize_aggregation_in_order
- read_in_order
- partition pruning for `COUNT()` optimization (`optimize_trivial_count_query=1`) ?

Reply:
Yes, in the first version I did that, but I found it may involve bugs if we simply replace the select expressions.

Eg: `00262_alter_alias.sql`.

So currently it only rewrites filter query, and just improves partition prune, indexes, alias on alias.

Reply:
>optimize_aggregation_in_order
>read_in_order

Then maybe these can be done in `ReadInOrderOptimizer::getInputOrder` manually ?

>partition pruning for COUNT() optimization (optimize_trivial_count_query=1) ?

query can be rewritten for completely for select count() optimization I guess (for simplifity), and in this case it should work (if I'm not missing anything)